File: infra/components/storage.py
"""GCS buckets for genomics data, ML models, and eval fixtures."""

import pulumi
import pulumi_gcp as gcp


class GCSBuckets(pulumi.ComponentResource):
    """Three GCS buckets: data, models (with lifecycle), and eval fixtures."""

    data_bucket_name: pulumi.Output[str]
    model_bucket_name: pulumi.Output[str]
    eval_fixtures_bucket_name: pulumi.Output[str]

    def __init__(
        self,
        name: str,
        project_id: str,
        region: str,
        opts: pulumi.ResourceOptions | None = None,
    ):
        super().__init__("genomics:infra:GCSBuckets", name, None, opts)

        child = pulumi.ResourceOptions(parent=self)

        self.data_bucket = gcp.storage.Bucket(
            f"{name}-data",
            name=f"{project_id}-genomics-data",
            project=project_id,
            location=region,
            uniform_bucket_level_access=True,
            force_destroy=False,
            versioning=gcp.storage.BucketVersioningArgs(enabled=True),
            opts=child,
        )

        self.model_bucket = gcp.storage.Bucket(
            f"{name}-models",
            name=f"{project_id}-genomics-models",
            project=project_id,
            location=region,
            uniform_bucket_level_access=True,
            force_destroy=False,
            versioning=gcp.storage.BucketVersioningArgs(enabled=True),
            lifecycle_rules=[
                gcp.storage.BucketLifecycleRuleArgs(
                    condition=gcp.storage.BucketLifecycleRuleConditionArgs(
                        num_newer_versions=5,
                    ),
                    action=gcp.storage.BucketLifecycleRuleActionArgs(type="Delete"),
                )
            ],
            opts=child,
        )

        self.eval_fixtures_bucket = gcp.storage.Bucket(
            f"{name}-eval-fixtures",
            name=f"{project_id}-genomics-eval-fixtures",
            project=project_id,
            location=region,
            uniform_bucket_level_access=True,
            force_destroy=False,
            opts=child,
        )

        self.data_bucket_name = self.data_bucket.name
        self.model_bucket_name = self.model_bucket.name
        self.eval_fixtures_bucket_name = self.eval_fixtures_bucket.name

        self.register_outputs(
            {
                "data_bucket_name": self.data_bucket_name,
                "model_bucket_name": self.model_bucket_name,
                "eval_fixtures_bucket_name": self.eval_fixtures_bucket_name,
            }
        )

File: infra-ts/components/storage.ts
import * as pulumi from "@pulumi/pulumi";
import * as gcp from "@pulumi/gcp";

export class GCSBuckets extends pulumi.ComponentResource {
  public readonly dataBucketName: pulumi.Output<string>;
  public readonly modelBucketName: pulumi.Output<string>;

  constructor(
    name: string,
    args: { projectId: string; region: string },
    opts?: pulumi.ComponentResourceOptions,
  ) {
    super("genomics:infra:GCSBuckets", name, {}, opts);
    const child = { parent: this };

    const dataBucket = new gcp.storage.Bucket(`${name}-data`, {
      name: pulumi.interpolate`${args.projectId}-genomics-data`,
      project: args.projectId,
      location: args.region,
      uniformBucketLevelAccess: true,
      versioning: { enabled: true },
      labels: { "data-classification": "phi", "hipaa-scope": "true" },
    }, child);

    const modelBucket = new gcp.storage.Bucket(`${name}-models`, {
      name: pulumi.interpolate`${args.projectId}-genomics-models`,
      project: args.projectId,
      location: args.region,
      uniformBucketLevelAccess: true,
      versioning: { enabled: true },
      labels: { "data-classification": "internal", "hipaa-scope": "false" },
    }, child);

    this.dataBucketName = dataBucket.name;
    this.modelBucketName = modelBucket.name;

    this.registerOutputs({
      dataBucketName: this.dataBucketName,
      modelBucketName: this.modelBucketName,
    });
  }
}
